### PR TITLE
Live example/knobs

### PR DESCRIPTION
**Pull request by shaneeza**

### CODE REVIEW

File: src/app/[component]/live-example/page.tsx [hunk 1]
```diff
@@ -38,23 +92,76 @@ export default function Page({ params }: { params: { component: string } }) {
...
-          data.allData?.default?.argTypes
-        )
-      );
+      const normalizedProps = createDefaultProps(data, darkMode);
+      setKnobProps(normalizedProps);
+
+      const propsWithValue: ComponentProps = {};
+      // creates an object with all the prop names and the values
+      for (let key in normalizedProps) {
+        // if (key === "children") {
```
Comment (bruugey) [inline src/app/[component]/live-example/page.tsx]: rm?

File: src/app/[component]/live-example/page.tsx [hunk 2]
```diff
@@ -38,23 +92,76 @@ export default function Page({ params }: { params: { component: string } }) {
...
+      setComponentProps(propsWithValue);
     }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
   }, [data]);
 
   useEffect(() => {
-    console.log({ props });
-  }, [props]);
+    console.log({ props: knobProps, componentProps });
+  }, [componentProps, knobProps]);
```
Comment (bruugey) [inline src/app/[component]/live-example/page.tsx]: rm?

File: src/app/[component]/live-example/page.tsx [hunk 3]
```diff
@@ -38,23 +92,76 @@ export default function Page({ params }: { params: { component: string } }) {
...
+
+  const updateKnobValue = (propName: string, newValue: any) => {
+    console.log("🐞updateKnobValue🐞", { propName, newValue });
```
Comment (bruugey) [inline src/app/[component]/live-example/page.tsx]: rm

File: src/app/[component]/live-example/page.tsx [hunk 4]
```diff
@@ -38,23 +92,76 @@ export default function Page({ params }: { params: { component: string } }) {
...
+          ...props[propName],
+          value: newValue,
+        },
+      };
+    });
+
+    setComponentProps((props) => {
+      return {
+        ...props,
+        [propName]: newValue, // TODO: will break if not a string or boolean
```
Comment (bruugey) [inline src/app/[component]/live-example/page.tsx]: is this still true?

File: src/components/live-example/Knob/Knob.styles.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import { css } from '@leafygreen-ui/emotion';
+
+export const radioBoxOverrideStyle = css`
+  & > div {
```
Comment (bruugey) [inline src/components/live-example/Knob/Knob.styles.ts]: we can inline this?

File: src/components/live-example/Knob/Knob.tsx [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+export const Knob = ({
+  propName,
+  knobType: knobTypeProp,
+  value,
+  onChange,
+  knobOptions,
+  ...rest
+}: KnobProps) => {
+  console.log("🪼🪼", { propName, value, knobTypeProp, knobOptions });
```
Comment (bruugey) [inline src/components/live-example/Knob/Knob.tsx]: rm

File: src/components/live-example/Knob/Knob.tsx [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+          onChange={onChange}
+          className={cx(inputStyle, rest.className)}
+          min={(knobTypeProp as KnobTypeObj)?.min ?? undefined}
+          max={(knobTypeProp as KnobTypeObj)?.max ?? undefined}
+        />
+      );
+
+    case "boolean":
+      return (
+        /// @ts-expect-error
```
Comment (bruugey) [inline src/components/live-example/Knob/Knob.tsx]: why expect error? i think you just need to pass an aria-labelledby

File: src/components/live-example/Knob/Knob.tsx [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+      );
+
+    case "array":
+    case "enum":
+    case "select":
+    case "radio": {
+      if (knobOptions && knobOptions.length) {
+        if (knobOptions.length <= 3) {
+          return (
+            <RadioBoxGroup
```
Comment (bruugey) [inline src/components/live-example/Knob/Knob.tsx]: do we want this or should we just render a select no matter what? ( i think the radio boxes are a little weird haha)

File: src/components/live-example/Knob/Knob.tsx [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+                >
+                  {opt}
+                </RadioBox>
+              ))}
+            </RadioBoxGroup>
+          );
+        }
+
+        return (
+          /// @ts-expect-error
```
Comment (bruugey) [inline src/components/live-example/Knob/Knob.tsx]: ditto here just add an aria-label

File: src/components/live-example/Knob/RawKnob.tsx [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+  }
+`;
+
+/**
+ * Enables users to edit & input valid JSON as a knob.
+ *
+ * A user can click the `edit` button to edit it the value
+ * And then click `save` to set the value in the example
+ */
+export const RawKnob = ({
```
Comment (bruugey) [inline src/components/live-example/Knob/RawKnob.tsx]: When do we use Raw Knobs?

File: src/components/live-example/Knob/RawKnob.tsx [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+      onChange(json);
+      setEditing(false);
+    } catch (err) {
+      setError(true);
+    }
+  };
+
+  return editing ? (
+    <div className={editModeWrapperStyle}>
+      {/* @ts-expect-error */}
```
Comment (bruugey) [inline src/components/live-example/Knob/RawKnob.tsx]: aria-label

File: src/components/live-example/KnobRow.tsx [hunk 12]
```diff
@@ -0,0 +1,91 @@
+import { kebabCase } from "lodash";
+// import { isRequired } from 'utils/tsdoc.utils';
```
Comment (bruugey) [inline src/components/live-example/KnobRow.tsx]: rm

File: src/components/live-example/KnobRow.tsx [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+import { spacing } from "@leafygreen-ui/tokens";
+import Tooltip from "@leafygreen-ui/tooltip";
+import { Body } from "@leafygreen-ui/typography";
+
+import { Knob } from "./Knob/Knob";
+
+const knobRowWrapperStyle = (darkMode: boolean) => css`
+  display: flex;
+  width: 100%;
+  min-height: ${spacing[5]}px;
```
Comment (bruugey) [inline src/components/live-example/KnobRow.tsx]: update to use our new spacing tokens

File: src/components/live-example/KnobRow.tsx [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+  justify-content: space-between;
+  align-items: center;
+  padding: ${spacing[4]}px;
+  gap: ${spacing[2]}px;
+  border-top: 1px solid ${darkMode ? palette.gray.dark2 : palette.gray.light2};
```
Comment (bruugey) [inline src/components/live-example/KnobRow.tsx]: change to use color tokens

File: src/components/live-example/KnobRow.tsx [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+    margin: unset;
+  }
+`;
+
+const knobControlStyle = css`
+  display: inline-flex;
+  justify-content: end;
+`;
+
+// const requiredFlagStyle = css`
```
Comment (bruugey) [inline src/components/live-example/KnobRow.tsx]: rm

File: src/components/live-example/KnobRow.tsx [hunk 16]
```diff
@@ -0,0 +1,91 @@
...
+//   display: inline;
+//   padding-left: 1ch;
+//   color: ${palette.red.base};
+//   text-transform: uppercase;
+// `;
+
+interface KnobRowProps extends HTMLElementProps<"div"> {
+  // knob: KnobType;
```
Comment (bruugey) [inline src/components/live-example/KnobRow.tsx]: rm

File: src/components/live-example/KnobRow.tsx [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+  return (
+    <div className={knobRowWrapperStyle(darkMode)}>
+      <div>
+        <Body
+          baseFontSize={16}
+          darkMode={darkMode}
+          id={`${kebabCase()}-knob-${name}`}
+        >
+          <strong>{name}</strong>
+          {/* {isRequired(knob) && (
```
Comment (bruugey) [inline src/components/live-example/KnobRow.tsx]: rm

File: src/components/live-example/Knob/Knob.tsx (hunk 8 above)
Comment (shaneeza) [reply src/components/live-example/Knob/Knob.tsx]: i think we can render a select

File: src/components/live-example/Knob/RawKnob.tsx (hunk 10 above)
Comment (shaneeza) [reply src/components/live-example/Knob/RawKnob.tsx]: I copied this from .design, but I don't think we need this anymore since we are omitting props.

**Pull request merged**